hadoop-hdds/tools/src/test/java/org/apache/hadoop/hdds/scm/cli/container/upgrade/TestUpgradeManager.java

```diff
@@ -0,0 +1,312 @@
+    datanodeId = UUID.randomUUID();
+    volumeSet = new MutableVolumeSet(datanodeId.toString(), scmId, CONF,
+        null, StorageVolume.VolumeType.DATA_VOLUME, null);
+
+    // create rocksdb instance in volume dir
+    final List<HddsVolume> volumes = new ArrayList<>();
+    for (StorageVolume storageVolume : volumeSet.getVolumesList()) {
+      HddsVolume hddsVolume = (HddsVolume) storageVolume;
+      StorageVolumeUtil.checkVolume(hddsVolume, scmId, scmId, CONF, null, null);
+      volumes.add(hddsVolume);
+    }
+
+    DatanodeDetails datanodeDetails = Mockito.mock(DatanodeDetails.class);
+    Mockito.when(datanodeDetails.getUuidString())
+        .thenReturn(datanodeId.toString());
+    Mockito.when(datanodeDetails.getUuid()).thenReturn(datanodeId);
+
+    volumeChoosingPolicy = mock(RoundRobinVolumeChoosingPolicy.class);
+    final AtomicInteger loopCount = new AtomicInteger(0);
+    Mockito.when(volumeChoosingPolicy.chooseVolume(anyList(), anyLong()))
+        .thenAnswer(invocation -> {
+          final int ii = loopCount.getAndIncrement() % volumes.size();
+          return volumes.get(ii);
+        });
+
+    containerSet = new ContainerSet(1000);
+
+    blockManager = new BlockManagerImpl(CONF);
+    chunkManager = new FilePerBlockStrategy(true, blockManager, null);
+  }
+
+  @After
+  public void after() throws IOException {
+    FileUtils.deleteDirectory(testRoot);
+  }
+
+  @Test
+  public void testUpgrade() throws IOException {
+    int num = 2;
+
+    final Map<KeyValueContainerData, Map<String, BlockData>>
+        keyValueContainerBlockDataMap = genSchemaV2Containers(num);
+    assertEquals(num, keyValueContainerBlockDataMap.size());
+
+    shutdownAllVolume();
+
+    final UpgradeManager upgradeManager = new UpgradeManager();
+    upgradeManager.initVolumeStoreMap(volumeSet, CONF);
+    final List<UpgradeManager.Result> results =
+        upgradeManager.upgradeAll(volumeSet, CONF);
+
+    checkV3MetaData(keyValueContainerBlockDataMap, results, upgradeManager);
+  }
+
+  private Map<String, BlockData> putAnyBlockData(KeyValueContainerData data,
+                                                 KeyValueContainer container,
+                                                 int numBlocks) {
+    // v3 key ==> block data
+    final Map<String, BlockData> containerBlockDataMap = new HashMap<>();
+
+    int txnID = 0;
+    for (int i = 0; i < numBlocks; i++) {
+      txnID = txnID + 1;
+      BlockID blockID =
+          ContainerTestHelper.getTestBlockID(data.getContainerID());
+      BlockData kd = new BlockData(blockID);
+      List<ContainerProtos.ChunkInfo> chunks = Lists.newArrayList();
+      putChunksInBlock(1, i, chunks, container, blockID);
+      kd.setChunks(chunks);
+
+      try {
+        final String localIDKey = Long.toString(blockID.getLocalID());
+        final String blockKey = DatanodeSchemaThreeDBDefinition
+            .getContainerKeyPrefix(data.getContainerID()) + localIDKey;
```

Comment:
You can try 

    final String blockKey = data.getBlockKey(blockID.getLocalID());

Reply:
> You can try
`final String blockKey = data.getBlockKey(blockID.getLocalID()); `


The ContainerData object Schema represented by data is v2, so I'll generate the key directly using the v3 approach

Reply:
I see.